### PR TITLE
Make Annotation's localPosition relative to object

### PR DESCRIPTION
**Pull request by dkwingsmt**
## Description

This PR changes how `AnnotatedRegionLayer` calculates `AnnotationEntry.localPosition` from the position relative to the layer to the position relative to the annotation object. 

Using the position relative to the layer was a design mistake. The offset of the layer is meaningless, and is unreliable. For example, a `Transform` widget will draw on the same layer with an offset if the transform matrix is a simple translation, or push a dedicated `TransformLayer` if the matrix is non-trivial. The former case will keep the previous coordinate origin (for example, the top left corner of the app), while the latter case will move the position origin since it's on a new layer. 

Moreover, the major (if not only) use case of `AnnotatedRegionLayer` is for render objects or widget to mark a region. It is expected that the `localPosition` is the position relative to the region, which is represented in `AnnotatedRegionLayer` by `offset` and `size`, while how the search is implemented under the hood should be transparent.

This change is also a first step of [the "annotation tree" project](https://github.com/flutter/flutter/issues/49568), which plans to move the annotations from the layer tree into an independent tree. As a result, they will no longer have access to their relative position to any layers.

## Related Issues

- Part of https://github.com/flutter/flutter/issues/49568
- Prerequisite to https://github.com/flutter/flutter/issues/33675

## Tests

I _changed_ the following tests:

* AnnotatedRegionLayer.findAllAnnotations should clip its annotation using size and offset (positive)

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] Yes, this is a breaking change.
   - [x] I wrote a design doc: https://flutter.dev/go/base-annotations-local-position-on-region
   - [x] I got input from the developer relations team, specifically from: sfshaza2
   - [x] I wrote a migration guide: https://github.com/flutter/website/pull/3673 ( https://flutter.dev/docs/release/breaking-changes/annotations-return-local-position-relative-to-object )

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

**Comment by goderbauer** [comment]
(PR triage): We are waiting for feedback before merging because of breaking change.

**Pull request merged**
